packages/browser/src/lib/on-page-leave.ts

```diff
@@ -0,0 +1,27 @@
+/**
+ * Register event listener on document that fires when:
+ * * tab loses focus or tab close (in mobile or desktop)
+ * * click external link
+ * * click relative link (same domain)
+ * * soft refresh / hard refresh
+ *
+ * adapted from https://stackoverflow.com/questions/3239834/window-onbeforeunload-not-working-on-the-ipad/52864508#52864508,
+ */
+export const onPageLeave = (cb: (...args: unknown[]) => void) => {
+  let unloaded = false // prevents double firing if both are supported
+
+  // using document instead of window because of bug affecting browsers before safari 14 (detail in footnotes https://caniuse.com/?search=visibilitychange)
+  document.addEventListener('pagehide', () => {
+    if (unloaded) return
+    unloaded = true
+    cb()
+  })
+
+  document.addEventListener('visibilitychange', () => {
```

Comment:
Should we reset the `unloaded` back to false if the visibilityState is not `hidden`? That way if someone comes back to the tab we still support the same unload behaviour if they close later on?

Reply:
> Should we reset the unloaded back to false if the visibilityState is not hidden? That way if someone comes back to the tab we still support the same unload behaviour if they close later on?This is a good call out!

> Edit: Not sure if that also means we need to rehydrate the events and put them in the queue to ensure they're sent if we return to the existing tab, or if a.js will already just try to resend them.

This is a good point --  this PR is focused on the batched-dispatcher, which leaves the scenario of events in the main event queue -- we don't want to save to localStorage on tab away since everything (including unsent events in the queue) should be suspended in-memory anyway.

Reply:
Ah awesome, that makes sense then :phew: